src/testharness/TestHarness.cpp

```diff
@@ -51,27 +57,89 @@ void TestHarness::spawnThreads() {
+            flattenedList.push_back(std::ref(test));
+            exeList.push_back(exePair.first);
+            tcList.push_back(tcPair.first);
           }
-
-          // spawn a new thread executing its tests
-          std::thread t(&TestHarness::threadRunTestsForToolChain, this, tcPair.first, exePair.first, std::ref(subpackage.second));
-          threadPool.push_back(std::move(t));
-          numThreads++;
         }
       }
     }
   }
 
-  // Join any stragglers
-  assert(threadPool.size() <= (size_t) cfg.getNumThreads());
-  for (size_t i = 0; i < threadPool.size(); i++) {
-    threadPool[i].join();
+  // Let the load balancing be the responsability of the threads, instead of the supervisor
+  for (int64_t i = 0; i < cfg.getNumThreads(); i++) {
+    std::thread t(&TestHarness::threadRunTestBatch, this,
+                  std::ref(tcList),
+                  std::ref(exeList),
+                  std::ref(flattenedList),
+                  std::ref(currentIndex),
+                  std::ref(currentIndexLock));
+    threadPool.push_back(std::move(t));
+  }
+
+  // KILL ALL THE CHILDREN
+  for (auto& thread : threadPool) thread.join();
+}
+
+void TestHarness::threadRunTestBatch(std::reference_wrapper<std::vector<std::string>> toolchains,
+                        std::reference_wrapper<std::vector<std::string>> executables,
+                        std::reference_wrapper<std::vector<std::reference_wrapper<TestPair>>> tests,
+                        std::reference_wrapper<size_t> currentIndex,
+                        std::reference_wrapper<std::mutex> currentIndexLock)
+{
+  // Loop while we have not exhausted the available tests
+  while (currentIndex < tests.get().size()) {
+    std::cout << "Entering loop" << std::endl;
+    size_t tmpIndex;
+
+    {
+      // Lock the index
+      std::lock_guard<std::mutex> lock(currentIndexLock);
+      std::cout << "Mutex Locked" << std::endl;
+
+      // store the current index
+      tmpIndex = currentIndex;
+      // increment for the next lad
+      currentIndex += cfg.getBatchSize();
+    }
+    std::cout << "Mutex Unlocked" << std::endl;
+    std::cout << "Current index: " << currentIndex << " || Tests size: " << tests.get().size() << std::endl;
+
+    size_t endIndex = ((tmpIndex + cfg.getBatchSize()) >= tests.get().size()) ?  tests.get().size() : tmpIndex + cfg.getBatchSize();
+
+    std::cout << "Thread " << std::this_thread::get_id() << ": Executing tests [" << tmpIndex << ", " << endIndex << "]" << std::endl;
+
+    threadRunTestsForToolChain(std::ref(toolchains), std::ref(executables), std::ref(tests), tmpIndex, endIndex);
+  }
+}
+
+void TestHarness::threadRunTestsForToolChain(std::reference_wrapper<std::vector<std::string>> tcNames,
+                                             std::reference_wrapper<std::vector<std::string>> exeNames,
+                                             std::reference_wrapper<std::vector<std::reference_wrapper<TestPair>>> tests,
```

Comment:
The internal one here is correct. It should be `std::vector<std::reference_wrapper<TestPair>>&`.